packages/drivers/odsp-driver/src/odspDocumentDeltaConnection.ts

```diff
@@ -664,28 +664,28 @@ export class OdspDocumentDeltaConnection extends DocumentDeltaConnection {
 				super.addTrackedListener(
 					event,
 					(msg: ISignalMessage | ISignalMessage[], documentId?: string) => {
-						const msgs = Array.isArray(msg) ? msg : [msg];
-						if (!this.enableMultiplexing || !documentId) {
-							if (!documentId) {
-								assert(
-									msgs.every((m) => !m.targetClientId),
-									"targetClientId defined while documentId is not",
-								);
-							}
+						if (!this.enableMultiplexing) {
 							listener(msg, documentId);
 							return;
 						}
 
+						assert(
+							documentId !== undefined,
+							"documentId is required when multiplexing is enabled.",
+						);
+
 						if (documentId !== this.documentId) {
 							return;
 						}
 
+						const msgs = Array.isArray(msg) ? msg : [msg];
+
 						const filteredMsgs = msgs.filter(
 							(m) => !m.targetClientId || m.targetClientId === this.clientId,
 						);
 
 						if (filteredMsgs.length > 0) {
-							listener(filteredMsgs, documentId);
+							listener(filteredMsgs.length === 1 ? filteredMsgs[0] : filteredMsgs, documentId);
 						}
```

Comment:
Why was this complication added?

Reply:
Signal layer compat tests fails without this, specifically the layer version combination where you have a very old loader and the most recent driver layer. Old loader doesn't handle batched signals (ISignalMessage[]) so only individual ISignalMessage's should be passed when there's one element for backcompat,

Reply:
@WillieHabi, just going over some old PRs. This is a place where it would be very good to have a comment. Please add when you can.